src/qilib/utils/storage/interface.py

```diff
@@ -21,11 +21,41 @@
 import logging
 from abc import ABC, abstractmethod
 from datetime import datetime
-from typing import Any, Callable, Optional, Union
+from collections.abc import Sequence as SequenceBaseClass
+from typing import Any, Callable, Optional, Union, Sequence
 
 from qilib.utils.type_aliases import TagType
 
 
+class _LazySequence(SequenceBaseClass):  # type: ignore
+    def __init__(self, length: int, item_getter: Callable[[int], Any]):
+        """ Convert a length and method to retrieve an indexed item into a sequence with lazy evaluation
+
+        Args:
+            length: Length of the sequence to be represented
+            item_getter: Method to retrieve an item at the specified index
+        """
+        super(SequenceBaseClass, self).__init__()
+        self._length = length
+        self._item_getter = item_getter
+
+    def __repr__(self) -> str:
+        classname = ".".join([self.__module__, self.__class__.__qualname__])
+        return f'<{classname} at %x{id(self)} length {self._length} >'
+
+    def __len__(self) -> int:
+        r = self._length
+        return r
+
+    def __getitem__(self, index: Union[int, slice]) -> Any:
+        if isinstance(index, slice):
+            slice_range = range(index.start or 0, index.stop or len(self), index.step or 1)
+            return [self._item_getter(i) for i in slice_range]
```

Comment:
Here you return an actual list with the components in it. This somewhat defeats the purpose of having lazy evaluation. You could return a generator expression instead:
```
return (self._item_getter(i) for i in slice_range)
```